### PR TITLE
APP-113: [C] Cross-compilation of the notification demo

### PR DESCRIPTION
**Pull request by RostakaGmfun**
/cc @Slash32 @Dubland

### CODE REVIEW

File: notificationdemo/source/c/build.sh [hunk 1]
```diff
@@ -15,35 +15,100 @@
...
 fi
 
-APP_NAME="demo_client"
 PROJECT_HOME=$(pwd)
 BUILD_DIR="build"
+KAA_LIB_PATH="./libs/kaa"
+KAA_C_LIB_SRC_PATH="$KAA_LIB_PATH/src"
+KAA_SDK_TAR="kaa-c*.tar.gz"
 KAA_TOOLCHAIN_PATH_SDK=""
+# TODO: comment about inconvenience between KAA_TARGET and KAA_PLATFORM
```
Comment (rasendubi) [inline notificationdemo/source/c/build.sh]: Please, address this TODO

File: notificationdemo/source/c/build.sh [hunk 2]
```diff
@@ -15,35 +15,100 @@
...
+KAA_TARGET=
+KAA_PRODUCE_BINARY=
+KAA_REQUIRE_CREDENTIALS=
+
+select_arch() {
+    echo "Please enter a target (default is posix):"
```
Comment (rasendubi) [inline notificationdemo/source/c/build.sh]: Maybe display possible options?

File: notificationdemo/source/c/build.sh [hunk 3]
```diff
@@ -15,35 +15,100 @@
...
+    read target
+
+    # TODO: better case handling
+    case "$target" in
+    posix)
```
Comment (rasendubi) [inline notificationdemo/source/c/build.sh]: `posix|"")`

File: notificationdemo/source/c/build.sh [hunk 4]
```diff
@@ -15,35 +15,100 @@
...
+    SSID=
+    PASSWORD=
+
+    cd "$PROJECT_HOME"
     mkdir -p "$PROJECT_HOME/$BUILD_DIR"
     cd $BUILD_DIR
-    cmake -DWITH_EXTENSION_EVENT=OFF \
-          -DWITH_EXTENSION_CONFIGURATION=OFF \
+
+    if [ $KAA_REQUIRE_CREDENTIALS ]
```
Comment (rasendubi) [inline notificationdemo/source/c/build.sh]: Use `if ...; then`. See our [code style guide](https://github.com/kaaproject/kaa/blob/develop/doc/Customization-guide/Code-style/Shell/index.md)

File: notificationdemo/source/c/build.sh [hunk 5]
```diff
@@ -15,35 +15,100 @@
...
+    then
+        echo "Enter WiFi SSID:"
+        read SSID
+        echo "Enter WiFi Password:"
+        read PASSWORD
+    fi
+
+    # TODO: comments about KAA_PLATFORM and KAA_TARGET
```
Comment (rasendubi) [inline notificationdemo/source/c/build.sh]: TODO

File: notificationdemo/source/c/src/kaa_demo.c [hunk 6]
```diff
@@ -48,79 +42,108 @@ static const char *color_to_str(enum color code_color)
     }
 }
 
-void on_notification(void *context, uint64_t *topic_id, kaa_notification_t *notification)
+#define INPUT_BUFFER_LENGTH 8
+
+static int read_topic_id()
```
Comment (rasendubi) [inline notificationdemo/source/c/src/kaa_demo.c]: `read_topic_id(void)`

File: notificationdemo/source/c/src/kaa_demo.c [hunk 7]
```diff
@@ -48,79 +42,108 @@ static const char *color_to_str(enum color code_color)
...
 {
-    KAA_DEMO_UNUSED(context);
+    char input[INPUT_BUFFER_LENGTH];
+    int input_length = 0;
+    do {
+        char c = getchar();
+        if (c == '\n') {
+            input[input_length] = 0;
```
Comment (rasendubi) [inline notificationdemo/source/c/src/kaa_demo.c]: Use `'\0'` for chars

File: notificationdemo/source/c/src/kaa_demo.c [hunk 8]
```diff
@@ -48,79 +42,108 @@ static const char *color_to_str(enum color code_color)
...
+
+static void on_notification(void *context, uint64_t *topic_id, kaa_notification_t *notification)
+{
+    (void)context;
     if (notification->alert_message) {
             kaa_string_t *message = (kaa_string_t *)notification->alert_message;
-            printf("Notification for topic id '%lu' received\n", *topic_id);
-            printf("Notification body: %s\n", message->data);
-            printf("Message alert type: %s\n", color_to_str(notification->alert_type));
+            demo_printf("Notification for topic id '%lu' received\r\n", *topic_id);
```
Comment (rasendubi) [inline notificationdemo/source/c/src/kaa_demo.c]: Code style: too much indent

File: notificationdemo/source/c/src/kaa_demo.c [hunk 9]
```diff
@@ -48,79 +42,108 @@ static const char *color_to_str(enum color code_color)
...
+            demo_printf("Notification body: %s\r\n", message->data);
+            demo_printf("Message alert type: %s\r\n", color_to_str(notification->alert_type));
     } else {
-        printf("Error:Received notification's body is null\n");
+        demo_printf("Error:Received notification's body is null\r\n");
```
Comment (rasendubi) [inline notificationdemo/source/c/src/kaa_demo.c]: Add space after `:`

File: notificationdemo/source/c/src/kaa_demo.c [hunk 10]
```diff
@@ -48,79 +42,108 @@ static const char *color_to_str(enum color code_color)
...
-        printf("Failed to sync subscriptions\n");
+        demo_printf("Failed to sync subscriptions\r\n");
     }
 }
 
-int main(/*int argc, char *argv[]*/)
+int main(void)
 {
-    printf("Notification demo started\n");
+    /**
```
Comment (rasendubi) [inline notificationdemo/source/c/src/kaa_demo.c]: Shouldn't be doxygen (doxygen doesn't work for local variables)

File: common/c/targets/esp8266/driver/uart.c [hunk 11]
```diff
@@ -293,3 +293,11 @@ uart_init_new(void)
     */
 
 }
+
+int uart_getchar(void)
+{
+    // Wait until data arrives to RX FIFO
+    while(!((READ_PERI_REG(UART_STATUS(UART0))>>UART_RXFIFO_CNT_S) & UART_RXFIFO_CNT));
```
Comment (rasendubi) [inline common/c/targets/esp8266/driver/uart.c]: code style (add space after `while`)

Comment (rasendubi) [inline common/c/targets/esp8266/driver/uart.c]: and spaces around `>>`

File: notificationdemo/source/c/build.sh (hunk 1 above)
Comment (RostakaGmfun) [reply notificationdemo/source/c/build.sh]: I propose to remove this TODO since the `build.sh` scripts will be eliminated ASAP

**Comment by Slash32** [comment]
LGTM

**Comment by rasendubi** [comment]
LGTM, but please use ShellCheck next time. There are a couple of issues that could cause script failure.

File: notificationdemo/source/c/src/kaa_demo.c [hunk 12]
```diff
@@ -48,79 +42,108 @@ static const char *color_to_str(enum color code_color)
...
+    do {
+        char c = getchar();
+        if (c == '\n') {
+            input[input_length] = '\0';
+            break;
+        } else {
+            input[input_length] = c;
+            input_length++;
+        }
+    } while (input_length < INPUT_BUFFER_LENGTH);
```
Comment (rasendubi) [inline notificationdemo/source/c/src/kaa_demo.c]: non-terminated string in case of long input

**Comment by rasendubi** [comment]
Well, there seem to be a bug. NOT LGTM

File: notificationdemo/source/c/build.sh (hunk 1 above)
Comment (rasendubi) [reply notificationdemo/source/c/build.sh]: Addressing the TODO is ~5 minutes, and ASAP never happens soon enough (we release 0.10 with build.sh)

File: notificationdemo/source/c/src/kaa_demo.c (hunk 12 above)
Comment (RostakaGmfun) [reply notificationdemo/source/c/src/kaa_demo.c]: Uh, oh, thanks, fixed

**Comment by rasendubi** [comment]
LGTM

**Pull request merged**